Zend/zend_attributes.c

```diff
@@ -0,0 +1,43 @@
+#include "zend.h"
+#include "zend_API.h"
+#include "zend_attributes.h"
+
+void zend_attribute_validate_phpattribute(zval *attribute, int target)
+{
+	if (target != ZEND_ATTRIBUTE_TARGET_CLASS) {
+		zend_error(E_COMPILE_ERROR, "Only classes can be marked with <<PhpAttribute>>");
+	}
+}
+
+void zend_attribute_validate_phpcompilerattribute(zval *attribute, int target)
+{
+	zend_error(E_COMPILE_ERROR, "The PhpCompilerAttribute can only be used by internal classes, use PhpAttribute instead");
+}
+
+void zend_register_attribute_ce(void)
+{
+	zend_hash_init(&zend_attributes_internal_validators, 8, NULL, NULL, 1);
+
+	zend_class_entry ce;
+
+	INIT_CLASS_ENTRY(ce, "PhpAttribute", NULL);
+	zend_ce_php_attribute = zend_register_internal_class(&ce);
+	zend_ce_php_attribute->ce_flags |= ZEND_ACC_FINAL;
+
+	zend_compiler_attribute_register(zend_ce_php_attribute, zend_attribute_validate_phpattribute);
```

Comment:
@nikic as per discussion with @beberlei through emails, he mentioned the following:
> ...while testing more it appeared that a dedicated class for `PhpCompilerAttribute` will cause some self-inflicted problems, which is why we will probably unify this to a single class.

Reply:
@nikic there is no support for attributes on internal classes at the moment. I didn't think it was necessary.

Reply:
I do think we want to support internal classes/functions as well. Attributes on internal classes are needed to define internal attributes, otherwise we couldn't place `<<PhpAttribute>>` on the class. Attributes on internal functions would be needed to support `<<Deprecated>>` there. We currently use a function flag for that, but I think it would be nice to handle that through an attribute (that could include a customized message) in the future.

Reply:
@nikic If this is the only reason for adding support for it, then its not really needed imho, because you can always check for the existance of a validator for the class as a proxy for "is this an internal attribute?" and then Return `<<PhpAttribute>>` for `<<Deprecated>>` (example) based on that information.

of course if a framework is written in C and wants to use attributes (for whatever reason) then it couldn't with the current technical basis. It just feels so wasteful :)

Reply:
@beberlei I think this will break down once we get to things like `<<Repeatable>>`. We would also have to emulate in reflection "as if" the internal attribute had those annotations. Otherwise the information would not be available externally.

It's also a question of storage. We have (barely) space for a ZEND_ACC_DEPRECATED bit, but we do not want to add a `const char *deprecation_message` to each function structure. While attributes are a very inefficient way to store that information, it's also a generic mechanism that only takes up that one pointer if unused, but can hold any amount of information once used.